Automattic/wp-calypso
# Fix: Jetpack product card prices (currency code) not loading in some cases.

#### Changes proposed in this Pull Request

This PR fixes a bug where the Jetpack product card prices are not loading and they are stuck in a loading state in the UI.  This occurs only in one particular edge-case that I am aware of, which is during the jetpack connect flow when creating a new account with magic links. (See testing instructions below to see exactly how to reproduce it)

The reason prices aren't loading is because the user's `currency_code` value is coming back as `null`.

#### Implementation notes:

The `currency_code` is retrieved as a byproduct of either of these three fetch actions:

- `PRODUCTS_LIST_RECEIVE`
- `PLANS_RECEIVE`
- `SITE_PLANS_FETCH_COMPLETED`

([code reference here](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/current-user/reducer.js#L78))

In this particular connection flow, none of the above three actions take place, therefore `currency_code` is coming back as `null`.

The fix was to add an additional fetch action, `SITE_PRODUCTS_FETCH_COMPLETED` because in this particular flow, this action is called and the `currency_code` can be retrieved from it.


#### Testing instructions

- Checkout this PR and start Calypso blue env (`yarn start`)

First, to reproduce and view the issue:

- Make sure you are logged-out of WordPress.com
- Create a JN site
- Go to https://wordpress.com/jetpack/connect and enter the JN site address, and click "Continue" button.
- On the log-in page, enter a new email address that is **not associated with an existing WordPress.com account**. (We'll be creating a new account here.) and click "Continue" button.
- Open your email account and open the "Create WordPress.com account" email that was sent to you. Click the button labeled, "Finish your Jetpack setup".
- Wait for the authorization process to finish...
- You should end up on the `wordpress.com/jetpack/connect/plans/:site?redirect=...` page.  You should see that the product prices are not loading and stuck in a loading state.
- You can refresh the page to confirm that the prices are still not loading.

Now to verify this PR fixes the issue:

- In the URL change the hostname from wordpress.com to calypso.localhost:3000 and press enter.
- Verify that all the prices in the product cards are now loading.

#### Testing steps screencast
![Screen Capture on 2021-06-28 at 15-57-29](https://user-images.githubusercontent.com/11078128/123696837-35eeca80-d82a-11eb-95a6-0ce4e845e2eb.gif)







Asana task # 1164141197617539-as-1200500830663033
